File: LintCode-Java/Triangle.java
import java.util.ArrayList;

public class Triangle {
    /**
     * @param triangle: a list of lists of integers.
     * @return: An integer, minimum path sum.
     */

    // V1, O(n^2)
    // DP(matrix, bottom up)
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0) {
            return 0;
        }

        int n = triangle.size();
        int[][] sum = new int[n][n];

        for (int i = 0; i < n; i++) {
            sum[n - 1][i] = triangle.get(n - 1).get(i);
        }

        for (int i = n - 2; i >= 0; i--) {
            for (int j = 0; j <= i; j++) {
                sum[i][j] = Math.min(sum[i + 1][j], sum[i + 1][j + 1]) + triangle.get(i).get(j);
            }
        }

        return sum[0][0];
    }

    // V2, O(2^n)
    // DFS(traverse)
    private int minSum = Integer.MAX_VALUE;
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0) {
            return 0;
        }
        // int minSum = Integer.MAX_VALUE;
        int n = triangle.size();
        traverse(0, 0, 0, triangle);
        return minSum;
    }
    private void traverse(int x, int y, int sum, ArrayList<ArrayList<Integer>> triangle) {
        if (x == triangle.size()) {
            minSum = Math.min(minSum, sum);
            return;
        }

        traverse(x + 1, y, sum + triangle.get(x).get(y), triangle);
        traverse(x + 1, y + 1, sum + triangle.get(x).get(y), triangle);
    }

    // V3, O(2^n)
    // Devide&Conquer
    // Could be improved to O(n^2) using hash storing tmp val
    public int minimumTotal(ArrayList<ArrayList<Integer>> triangle) {
        if (triangle == null || triangle.size() == 0) {
            return 0;
        }
        return helper(0, 0, triangle);
    }
    private int helper(int x, int y, ArrayList<ArrayList<Integer>> triangle) {
        if (x == triangle.size()) {
            return 0;
        }

        return triangle.get(x).get(y) + Math.min(helper(x + 1, y, triangle), helper(x + 1, y + 1, triangle));
    }
}

File: LintCode-Python/triangle.py
class Solution:
    """
    @param triangle: a list of lists of integers.
    @return: An integer, minimum path sum.
    """
    min_sum = sys.maxint
    def minimumTotal(self, triangle):
        # [dfs]
        # time:  O(n?)
        # space: O(1?)
        if not triangle or not triangle[0]:
            return 0
        self.traverse(triangle, 0, 0, 0)
        return self.min_sum

    def traverse(self, triangle, row, col, curr_sum):
        if row == len(triangle):
            self.min_sum = min(self.min_sum, curr_sum)
            return

        self.traverse(triangle, row + 1, col, curr_sum + triangle[row][col])
        self.traverse(triangle, row + 1, col + 1, curr_sum + triangle[row][col])

    def minimumTotal(self, triangle):
        # [divide and conquer]
        # time:  O(n?)
        # space: O(1?)
        if not triangle:
            return 0
        return self.helper(triangle, 0, 0)

    def helper(self, triangle, row, col):
        if row > len(triangle) - 1 or col > len(triangle[row]) - 1:
            return 0

        # Divide
        leftSum = self.helper(triangle, row + 1, col)
        rightSum = self.helper(triangle, row + 1, col + 1)

        # Conquer
        return min(leftSum, rightSum) + triangle[row][col]

